### PR TITLE
Chainlink CCIP Metrics

### PR DESCRIPTION
**Pull request by AnonJon**
Chainlink CCIP metrics including:
- tx daily count
- daily gas spend
- daily rewards

Chains:
- mainnet
- arbitrum
- optimism
- bnb
- polygon
- avax
- base

### CODE REVIEW

File: models/chainlink/arbitrum/chainlink_arbitrum_ccip_reverted_transactions.sql [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    SELECT
+      tx.tx_hash as tx_hash,
+      MAX(tx.block_time) as block_time,
+      cast(date_trunc('month', MAX(tx.block_time)) as date) as date_month,
+      tx."from" as "node_address",
+      MAX((cast((gas_used) as double) / 1e18) * effective_gas_price) as token_amount,
+      MAX(arbitrum_usd.usd_amount) as usd_amount
+    FROM
+      {{ ref('chainlink_arbitrum_ccip_send_traces') }} tx
+      LEFT JOIN {{ source('arbitrum', 'transactions') }} tx2 ON tx2.hash = tx.tx_hash
```
Comment (jeff-dude) [inline models/chainlink/arbitrum/chainlink_arbitrum_ccip_reverted_transactions.sql:38]: can you add incremental filter on this base table join condition?

also, if possible, an inner join would help performance

File: models/chainlink/arbitrum/chainlink_arbitrum_ccip_reward_daily.sql [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        FROM
+            {{ref('chainlink_arbitrum_ccip_token_meta')}}
+    ),
+    token_usd_daily AS (
+        SELECT
+            cast(date_trunc('day', price.minute) as date) as "date_start",
+            token_meta.token_symbol as symbol,
+            MAX(price.price) as usd_amount
+        FROM
+            {{ source('prices', 'usd') }} price
```
Comment (jeff-dude) [inline models/chainlink/arbitrum/chainlink_arbitrum_ccip_reward_daily.sql:28]: incremental filter?

File: models/chainlink/arbitrum/chainlink_arbitrum_ccip_reward_daily.sql [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    ),
+    ccip_reward_daily AS (
+        SELECT 
+            ccip_send_requested_daily.date_start,
+            cast(date_trunc('month', ccip_send_requested_daily.date_start) as date) as date_month,  
+            SUM(ccip_send_requested_daily.fee_amount) as token_amount,
+            SUM((ccip_send_requested_daily.fee_amount * tud.usd_amount)) as usd_amount,
+            ccip_send_requested_daily.token as token
+        FROM 
+            {{ref('chainlink_arbitrum_ccip_send_requested_daily')}} ccip_send_requested_daily
```
Comment (jeff-dude) [inline models/chainlink/arbitrum/chainlink_arbitrum_ccip_reward_daily.sql:43]: incremental filter?

File: models/chainlink/arbitrum/chainlink_arbitrum_ccip_transmitted_reverted.sql [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      tx.tx_hash as tx_hash,
+      tx.index as tx_index,
+      MAX(tx.block_time) as block_time,
+      cast(date_trunc('month', MAX(tx.block_time)) as date) as date_month,
+      tx.tx_from as "node_address",
+      MAX((cast((gas_used) as double) / 1e18) * effective_gas_price) as token_amount,
+      MAX(arbitrum_usd.usd_amount) as usd_amount
+    FROM
+      {{ ref('chainlink_arbitrum_ccip_transmitted_logs') }} tx
+      LEFT JOIN {{ source('arbitrum', 'transactions') }} tx2 ON tx2.hash = tx.tx_hash
```
Comment (jeff-dude) [inline models/chainlink/arbitrum/chainlink_arbitrum_ccip_transmitted_reverted.sql:39]: incremental filter?

**Review by jeff-dude — CHANGES_REQUESTED** [review]
few minor comments below on arbitrum only, applies universally

File: models/chainlink/arbitrum/chainlink_arbitrum_ccip_transmitted_reverted.sql (hunk 4 above)
Comment (AnonJon) [reply models/chainlink/arbitrum/chainlink_arbitrum_ccip_transmitted_reverted.sql:39]: Incremental filters are in on all networks now

**Review by jeff-dude — CHANGES_REQUESTED** [review]
looks like we need it cleaned up a bit more. i commented on two bnb models, since bnb is taking the longest in test runs, but the comments appear to apply universally again

i believe i mentioned this in another PR to you & team, but i think you'd really benefit from migrating this code to the new macro approach. that way, you'll have each model code in one place only (the macro) and then each blockchain spell can call that macro and pass in `blockchain` values as needed.

do you want to try and build that out in this PR for the new models? then once we get finalized, you can look into migrating all other chainlink spells into a similar macro approach?

examples:
https://github.com/duneanalytics/spellbook/pull/4936/files#diff-9fab187daefbc12cc7c9c9c77f8d44205c257dfcf12fc7954e0f56f9968d84c7

File: models/chainlink/bnb/chainlink_bnb_ccip_reverted_transactions.sql [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      MAX(tx.block_time) as block_time,
+      cast(date_trunc('month', MAX(tx.block_time)) as date) as date_month,
+      tx."from" as "node_address",
+      MAX((cast((gas_used) as double) / 1e18) * gas_price) as token_amount,
+      MAX(bnb_usd.usd_amount) as usd_amount
+    FROM
+      {{ ref('chainlink_bnb_ccip_send_traces') }} tx
+      INNER JOIN {{ source('bnb', 'transactions') }} tx2 ON tx2.hash = tx.tx_hash
+      {% if is_incremental() %}
+        AND tx.block_time >= date_trunc('day', now() - interval '{{incremental_interval}}' day)
```
[the remark this answers is not in the document]
Comment (AnonJon) [reply models/chainlink/bnb/chainlink_bnb_ccip_reverted_transactions.sql:40]: ah yes I see, ill update the rest of the networks.

File: models/chainlink/bnb/chainlink_bnb_ccip_request_daily.sql [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      COALESCE(COUNT(fulfilled.token_amount), 0) + COALESCE(COUNT(reverted.token_amount), 0) as total_requests
+    FROM
+      {{ ref('chainlink_bnb_ccip_fulfilled_transactions') }} fulfilled
+      FULL OUTER JOIN {{ ref('chainlink_bnb_ccip_reverted_transactions') }} reverted ON
+        reverted.block_time = fulfilled.block_time AND
+        reverted.node_address = fulfilled.node_address
+    {% if is_incremental() %}
+      WHERE
+        fulfilled.block_time >= date_trunc('day', now() - interval '{{incremental_interval}}' day)
+        OR reverted.block_time >= date_trunc('day', now() - interval '{{incremental_interval}}' day)
```
[the remark this answers is not in the document]
Comment (AnonJon) [reply models/chainlink/bnb/chainlink_bnb_ccip_request_daily.sql:34]: As for this I suppose it should be AND. Let me make the change and if it helps the incremental.

**Comment by AnonJon** [comment]
I did bring back the idea of the macro approach back to my team when you posted about it before. I cant remember that exact reason but we came to the conclusion that because of some differences in network needs and engineering time currently we were not going to go that route as of right now. The CCIP PR should be the last major Chainlink product we focus on making metrics for as well but next year if we free up some time we will reevaluate migrating over.

**Review by jeff-dude — APPROVED** [review]

**Comment by jeff-dude** [comment]
> I did bring back the idea of the macro approach back to my team when you posted about it before. I cant remember that exact reason but we came to the conclusion that because of some differences in network needs and engineering time currently we were not going to go that route as of right now. The CCIP PR should be the last major Chainlink product we focus on making metrics for as well but next year if we free up some time we will reevaluate migrating over.

that'd be much appreciated. thanks for the info

**Pull request merged**